lib/atoi/a2i.h

```diff
@@ -15,12 +15,76 @@
 #include "attr.h"
 
 
+#define a2i(TYPE, ...)                                                        \
+(                                                                             \
+	_Generic((TYPE) 0,                                                    \
+		short:              a2sh,                                     \
+		int:                a2si,                                     \
+		long:               a2sl,                                     \
+		long long:          a2sll,                                    \
+		unsigned short:     a2uh,                                     \
+		unsigned int:       a2ui,                                     \
+		unsigned long:      a2ul,                                     \
+		unsigned long long: a2ull                                     \
+	)(__VA_ARGS__)                                                        \
+)
+
+
+ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
+inline int a2sh(short *restrict n, const char *s,
+    char **restrict endp, int base, short min, short max);
+ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
+inline int a2si(int *restrict n, const char *s,
+    char **restrict endp, int base, int min, int max);
 ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
 inline int a2sl(long *restrict n, const char *s,
     char **restrict endp, int base, long min, long max);
 ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
+inline int a2sll(long long *restrict n, const char *s,
+    char **restrict endp, int base, long long min, long long max);
+ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
+inline int a2uh(unsigned short *restrict n, const char *s,
+    char **restrict endp, int base, unsigned short min, unsigned short max);
+ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
+inline int a2ui(unsigned int *restrict n, const char *s,
+    char **restrict endp, int base, unsigned int min, unsigned int max);
+ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
 inline int a2ul(unsigned long *restrict n, const char *s,
     char **restrict endp, int base, unsigned long min, unsigned long max);
+ATTR_STRING(2) ATTR_ACCESS(write_only, 1) ATTR_ACCESS(write_only, 3)
+inline int a2ull(unsigned long long *restrict n, const char *s,
+    char **restrict endp, int base, unsigned long long min,
+    unsigned long long max);
+
+
+inline int
+a2sh(short *restrict n, const char *s, char **restrict endp,
```

Comment:
All of these functions (including the existing ones in atoi/) need brief descriptions.

I'm ok merging this and waiting for a new PR that adds those.  On the other hand, having those would make this a lot easier to review.

Reply:
[a2i.3.pdf](https://github.com/shadow-maint/shadow/files/14590313/a2i.3.pdf)
[str2i.3.pdf](https://github.com/shadow-maint/shadow/files/14590314/str2i.3.pdf)

I've uploaded two PDFs of the manual pages I wrote for them in liba2i.  You can find the source here: <https://git.kernel.org/pub/scm/libs/liba2i/liba2i.git/tree/share/man>

Reply:
Do you want to add these man(7) pages to shadow?  However, they are written in man(7), while the existing pages in shadow are in XML.  I'm not sure if that's a problem.

Or maybe we could say that the documentation for these functions can be found in liba2i?

Reply:
My long-term plan (i.e., maybe for 2025), is to remove these definitions and add liba2i as a dependency.  But I need widespread support from distros, which will take time.

Reply:
Regarding the existing ones, they have a small description in the commits that added them.

strtoi_() and strtou_() are described to be identical to strtoi(3bsd) and strtou(3bsd), which are documented in the systems that provide them, so I guess we don't need to repeat that.  (And also except for a bug that existed when I added them, but that bug has already been fixed in both libbsd and NetBSD, fixed last January or so.)

<https://github.com/shadow-maint/shadow/commit/34ff8edb63f17032c6a215331249883a9bdf8cd1>

strtou_noneg() is described to be equivalent to strtou(), except that it rejects negative input.

<https://github.com/shadow-maint/shadow/commit/f2b240595b1c0f10f3a73cceb3bc4f6226cb8818>

If you think we need something else, please let me know.

Reply:
@hallyn 

> If you think we need something more, please let me know.

Ping ;)

Reply:
Sorry - yes, I do think more is needed.  There are way too many random seeming 6 character string manipulation functions with enormous implications for their proper use.  But we can improve that later.  Thanks.

Reply:
Hm, I'm confused:  taking a look at

 a2ul -> strtou_noneg -> strtoi_ -> strtou_noneg -> strtoi_(max = 1)

Won't that always get status == ERANGE and thence return min?

Reply:
I'm not sure what

> a2ul -> strtou_noneg -> strtoi_ -> strtou_noneg -> strtoi_(max = 1)

means.  Could you maybe explain that, or give some code example?

Reply:
Ahh, I guess you mean this:

```c
$ grepc -tfd a2ul .
./lib/atoi/a2i.h:inline int
a2ul(unsigned long *restrict n, const char *s, char **restrict endp,
    int base, unsigned long min, unsigned long max)
{
	int  status;

	*n = strtou_noneg(s, endp, base, min, max, &status);
	if (status != 0) {
		errno = status;
		return -1;
	}
	return 0;
}
```
It calls *strtou_noneg*(); let's search that:
```c
$ grepc -tfd strtou_noneg .
./lib/atoi/strtou_noneg.h:inline uintmax_t
strtou_noneg(const char *s, char **restrict endp, int base,
    uintmax_t min, uintmax_t max, int *restrict status)
{
	int  st;

	if (status == NULL)
		status = &st;
	if (strtoi_(s, endp, base, 0, 1, status) == 0 && *status == ERANGE)
		return min;

	return strtou_(s, endp, base, min, max, status);
}
```
It has the following:
```c
	if (strtoi_(s, endp, base, 0, 1, status) == 0 && *status == ERANGE)
		return min;
```
which seems to only accept 0 or 1.  But notice that it has a `== 0`, which means that we only take this branch if *strtoi_*() parsed a number <=0, and the following `&& *status == ERANGE` makes it so that only negative numbers trigger that branch.

When that *strtoi_*() call parses any positive number, it will return 1, so that the branch won't execute, and the following line will be called:
```c
	return strtou_(s, endp, base, min, max, status);
```

Reply:
Sorry - I meant that strtou_noneg() always passes 1 for the value
of 'max' when calling strtoi_().

Which it seems to me means that strtoi_() will set status to
ERANGE, and strtou_noneg() will see that and return min.

I should probably just set up a test program from the diff.
Because strtoN in particular may not be doing quite what I think.
Will see about doing that tomorrow.

Reply:
*strtou_noneg*() just calls *strtoi_*() for deciding if the input is negative.  It uses it as a boolean call, by abusing some parameters.

The actual work is in the following call, to *strtou_*().

Reply:
Thanks.

Reply:
There are currently tests for it, so unless the tests are broken, you can check them:

```c
alx@debian:~/src/shadow/shadow/master$ grepc -tfd strtou_noneg .
./lib/atoi/strtou_noneg.h:inline uintmax_t
strtou_noneg(const char *s, char **restrict endp, int base,
    uintmax_t min, uintmax_t max, int *restrict status)
{
	int  st;

	if (status == NULL)
		status = &st;
	if (strtoi_(s, endp, base, 0, 1, status) == 0 && *status == ERANGE)
		return min;

	return strtou_(s, endp, base, min, max, status);
}
alx@debian:~/src/shadow/shadow/master$ grep -rn strtou_noneg tests/
tests/unit/Makefile.am:9:    test_atoi_strtou_noneg \
tests/unit/Makefile.am:51:test_atoi_strtou_noneg_SOURCES = \
tests/unit/Makefile.am:52:    ../../lib/atoi/strtou_noneg.c \
tests/unit/Makefile.am:53:    test_atoi_strtou_noneg.c \
tests/unit/Makefile.am:55:test_atoi_strtou_noneg_CFLAGS = \
tests/unit/Makefile.am:58:test_atoi_strtou_noneg_LDFLAGS = \
tests/unit/Makefile.am:60:test_atoi_strtou_noneg_LDADD = \
tests/unit/test_atoi_strtoi.c:18:#include "atoi/strtou_noneg.h"
tests/unit/test_atoi_strtoi.c:23:static void test_strtou_noneg(void **state);
tests/unit/test_atoi_strtoi.c:32:        cmocka_unit_test(test_strtou_noneg),
tests/unit/test_atoi_strtoi.c:116:test_strtou_noneg(void **state)
tests/unit/test_atoi_strtoi.c:122:	assert_true(strtou_noneg("42", NULL, -1, 1, 2, &status)
tests/unit/test_atoi_strtoi.c:126:	assert_true(strtou_noneg("40", &end, 5, 0, UINTMAX_MAX, &status)
tests/unit/test_atoi_strtoi.c:131:	assert_true(strtou_noneg("-40", &end, 0, 2, UINTMAX_MAX, &status)
tests/unit/test_atoi_strtoi.c:136:	assert_true(strtou_noneg("z", &end, 0, 0, UINTMAX_MAX, &status)
tests/unit/test_atoi_strtoi.c:141:	assert_true(strtou_noneg(" 5", &end, 0, 3, 4, &status)
tests/unit/test_atoi_strtoi.c:146:	assert_true(strtou_noneg("5z", &end, 0, 0, UINTMAX_MAX, &status)
tests/unit/test_atoi_strtoi.c:151:	assert_true(strtou_noneg("5z", &end, 0, 0, 4, &status)
tests/unit/test_atoi_strtou_noneg.c:16:#include "atoi/strtou_noneg.h"
```